src/utils.jl

```diff
@@ -0,0 +1,249 @@
+# pre-1.11 compat
+if VERSION < v"1.11"
+    function _in!(x, s::Set)
+        xT = convert(eltype(s), x)
+        idx, sh = Base.ht_keyindex2_shorthash!(s.dict, xT)
+        idx > 0 && return true
+        Base._setindex!(s.dict, nothing, xT, -idx, sh)
+        return false
+    end
+else
+    const _in! = Base.in!
+end
+
+# hand-rolled scoped enum
+module JSONTypes
+    primitive type T 8 end
+    T(x::UInt8) = Base.bitcast(T, x)
+    const OBJECT = T(0x00)
+    const ARRAY = T(0x01)
+    const STRING = T(0x02)
+    const INT = T(0x03)
+    const FLOAT = T(0x04)
+    const FALSE = T(0x05)
+    const TRUE = T(0x06)
+    const NULL = T(0x07)
+    const NUMBER = T(0x08) # currently used by LazyValue
+    const names = Dict(
+        OBJECT => "OBJECT",
+        ARRAY => "ARRAY",
+        STRING => "STRING",
+        INT => "INT",
+        FALSE => "FALSE",
+        TRUE => "TRUE",
+        FLOAT => "FLOAT",
+        NULL => "NULL",
+        NUMBER => "NUMBER",
+    )
+    Base.show(io::IO, x::T) = Base.print(io, "JSONTypes.", names[x])
+end
+
+# windows doesn't like mmaped buffers hanging around and will error if you try to modify
+# so we call finalize eagerly to ensure the buffer is unmapped
+# ref: https://github.com/JuliaIO/JSON.jl/issues/112
+checkfinalize!(buf) = Sys.iswindows() && (VERSION ≥ v"1.11" ? finalize(buf.ref.mem) : finalize(buf))
```

Comment:
I think the `ref` field is not public.

Reply:
Good call. I'll make the compat here more defensive by also checking that `buf` has a `ref` field and that `buf.ref` has a `mem` field. The bigger picture issue here is that it was technically a breaking part of the move to `Memory` that mmapped arrays have the finalizer defined on the `ref.buf` instead of the array itself, so we're trying to deal w/ that breakage here in a way that still makes mmap usage ok for windows users.